File: grodb/src/nu/mine/mosher/gedcomold/GedcomLine.java
package nu.mine.mosher.gedcom;

public class GedcomLine
{
	private final int level;
	private final String id;
	private final String tag;
	private final String value;
	private final String pointer;
	public GedcomLine(int level, String id, String tag, String value)
	{
		this.level = level;
		this.id = getPointer(id);
		this.tag = tag;
		String v = getPointer(value);
		if (v.length() > 0)
		{
			this.pointer = v;
			this.value = "";
		}
		else
		{
			this.value = replaceAts(value);
			this.pointer = "";
		}
	}
	private static String getPointer(String s)
	{
		if (!s.startsWith("@") || !s.endsWith("@") || s.length() < 3)
		{
			return "";
		}

		String pointer = s.substring(1,s.length()-1);
		if (pointer.indexOf('@') >= 0)
		{
			return "";
		}
		return pointer;
	}
	private static String replaceAts(String s)
	{
		return s.replaceAll("@@","@");
	}
	public String toString()
	{
		StringBuffer sb = new StringBuffer(256);
		sb.append(level);
		sb.append(",");
		if (hasID())
		{
			sb.append("id=");
			sb.append(id);
			sb.append(",");
		}
		sb.append("tag=");
		sb.append(tag);
		sb.append(",");
		if (isPointer())
		{
			sb.append("pointer=");
			sb.append(pointer);
		}
		else
		{
			sb.append("value=\"");
			sb.append(value);
			sb.append("\"");
		}
		return sb.toString();
	}
	public boolean hasID()
	{
		return id.length() > 0;
	}
	public boolean isPointer()
	{
		return pointer.length() > 0;
	}
}

File: grojs/GedcomLine.js
/* PUBLIC */

/* constructor */

function GedcomLine(level, gid, tag, val) {
	Util.verifyType(this,"GedcomLine");

	if (!arguments.length) {
		return this;
	}

	this.level = parseInt(level);
	this.gid = GedcomLine.asPointer(gid);
	this.tag = Util.safeStr(tag).toUpperCase();
	this.pointer = GedcomLine.asPointer(val);
	if (this.pointer) {
		this.val = Util.safeStr();
	} else {
		this.val = GedcomLine.replaceAts(Util.safeStr(val));
	}
}

/* simple accessor methods */

GedcomLine.prototype.getLevel = function() {
	return this.level;
};

GedcomLine.prototype.getID = function() {
	return this.gid;
};

GedcomLine.prototype.getTag = function() {
	return this.tag;
};

GedcomLine.prototype.getVal = function() {
	return this.val;
};

GedcomLine.prototype.getPointer = function() {
	return this.pointer;
};

/* complex accessor methods */

GedcomLine.prototype.hasID = function() {
	return !!this.getID();
};

GedcomLine.prototype.isPointer = function() {
	return !!this.getPointer();
};

/* mutator methods */

GedcomLine.prototype.concat = function(c) {
	switch (c.getTag()) {
		case "CONC": this.val += c.getVal(); break;
		case "CONT": this.val += "\n"+c.getVal(); break;
	}
};

/* static factory methods */

GedcomLine.parse = function(s) {
	var r = /^(\d+)\s+(?:(@[^@]+@)\s+)?(\S+)(?:\s(.*))?$/.exec(s);
	if (r == null) {
		throw new Error("Gedcom line has invalid syntax: "+s);
	}
	return new GedcomLine(r[1],r[2],r[3],r[4]);
};


/* PRIVATE */

/* static methods */

GedcomLine.asPointer = function(s) {
	var r;
	r = /^@([^@]+)@$/.exec(s);
	if (r == null) {
		return "";
	}
	return r[1];
};

GedcomLine.replaceAts = function(s) {
	if (!s || !s.replace) {
		return s;
	}
	return s.replace(/@@/g,"@");
};
